README.md
# Rinvex Support

**Rinvex common** support helpers, contracts, and traits required by various Rinvex packages. Validator functionality, and basic controller included out-of-the-box.

[![Packagist](https://img.shields.io/packagist/v/rinvex/laravel-support.svg?label=Packagist&style=flat-square)](https://packagist.org/packages/rinvex/laravel-support)
[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/rinvex/laravel-support.svg?label=Scrutinizer&style=flat-square)](https://scrutinizer-ci.com/g/rinvex/laravel-support/)
[![Travis](https://img.shields.io/travis/rinvex/laravel-support.svg?label=TravisCI&style=flat-square)](https://travis-ci.org/rinvex/laravel-support)
[![StyleCI](https://styleci.io/repos/60968880/shield)](https://styleci.io/repos/60968880)
[![License](https://img.shields.io/packagist/l/rinvex/laravel-support.svg?label=License&style=flat-square)](https://github.com/rinvex/laravel-support/blob/develop/LICENSE)


> **Note:** this package is just a support package for other Rinvex packages, which may not be useful on it's own, but contains some complementary generic functionality and also may not respect SemVer and break backward compatibility.


## Installation

Install via `composer require rinvex/laravel-support`


## Usage

### `mimetypes()` helper

The `mimetypes` method gets valid mime types:
```php
$mimetypes = mimetypes();
```

### `timezones()` helper

The `timezones` method gets valid timezones:
```php
$timezones = timezones();
```

### unique_with Validator Rule

This feature contains a variant of the `validateUnique` rule for Laravel, that allows for validation of multi-column UNIQUE indexes.

It was forked and merged from the awesome [felixkiss/uniquewith-validator](https://github.com/felixkiss/uniquewith-validator) package, which at the time been outdated and un-maintained for a long time. Many thanks to core contributors for developing this.


## Changelog

Refer to the [Changelog](CHANGELOG.md) for a full history of the project.


## Support

The following support channels are available at your fingertips:

- [Chat on Slack](https://bit.ly/rinvex-slack)
- [Help on Email](mailto:help@rinvex.com)
- [Follow on Twitter](https://twitter.com/rinvex)


## Contributing & Protocols

Thank you for considering contributing to this project! The contribution guide can be found in [CONTRIBUTING.md](CONTRIBUTING.md).

Bug reports, feature requests, and pull requests are very welcome.

- [Versioning](CONTRIBUTING.md#versioning)
- [Pull Requests](CONTRIBUTING.md#pull-requests)
- [Coding Standards](CONTRIBUTING.md#coding-standards)
- [Feature Requests](CONTRIBUTING.md#feature-requests)
- [Git Flow](CONTRIBUTING.md#git-flow)


## Security Vulnerabilities

If you discover a security vulnerability within this project, please send an e-mail to [help@rinvex.com](help@rinvex.com). All security vulnerabilities will be promptly addressed.


## About Rinvex

Rinvex is a software solutions startup, specialized in integrated enterprise solutions for SMEs established in Alexandria, Egypt since June 2016. We believe that our drive The Value, The Reach, and The Impact is what differentiates us and unleash the endless possibilities of our philosophy through the power of software. We like to call it Innovation At The Speed Of Life. That’s how we do our share of advancing humanity.


## License

This software is released under [The MIT License (MIT)](LICENSE).

(c) 2016-2022 Rinvex LLC, Some rights reserved.


CHANGELOG.md
# Rinvex Support Change Log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](CONTRIBUTING.md).


## [v7.3.1] - 2023-08-30
- Support new accessarea file-based structure

## [v7.3.0] - 2023-08-28
- Support multiple themes
- Remove `/vendor` from namespaced view paths to simplify the path

## [v7.2.6] - 2023-08-16
- Register models as singletons by default

## [v7.2.5] - 2023-08-02
- If view global path exists in the view hints, remove and prepend to eliminate duplicates, and keep precedence
- Prepend view global paths to take precedence over modules & extensions

## [v7.2.4] - 2023-07-25
- Move unique_with validation rule to rinvex/laravel-support from cortex/foundation

## [v7.2.3] - 2023-07-22
- Remove intend() method docs to cortex/foundation
- Override ServiceProvider::loadViewsFrom to prepend namespaces instead of appending, allowing extensions to have precedence and override parent module views

## [v7.2.2] - 2023-07-03
- Update composer dependencies

## [v7.2.1] - 2023-06-29
- Refactor resource loading and publishing

## [v7.2.0] - 2023-06-10
- Add module extensions support to service provider vendor publishing
- Bind models in IoC service container only if not bound already
- Remove unused global function array_filter_recursive

## [v7.1.0] - 2023-05-02
- 4eee070: Add support for Laravel v11, and drop support for Laravel v9
- b0b7595: Upgrade spatie/laravel-schemaless-attributes to v2.4 from v2.3
- e97387d: Update watson/validating to v8.0 from v7.0
- 8543fdd: Update phpunit to v10.1 from v9.5

## [v7.0.1] - 2023-01-15
- Remove unused method call `runningInDevzone`

## [v7.0.0] - 2023-01-09
- Tweak artisan commands registration
- Drop PHP v8.0 support and update composer dependencies

## [v6.1.5] - 2022-12-30
- Drop felixkiss/uniquewith-validator usage
  - This is no longer a requirement, and could be included on the project level if needed. 

## [v6.1.4] - 2022-12-30
- Drop `hashid` model attribute append to fix toArray issues in some scenarios, we can always call it explicitly if needed.

## [v6.1.3] - 2022-08-30
- Tweak hashids trait

## [v6.1.2] - 2022-08-06
- Tweak model IDs hashing to support configurable hashed keys

## [v6.1.1] - 2022-06-11
- Fix null string issue with strip_tags

## [v6.1.0] - 2022-02-14
- Update composer dependencies to Laravel v9
- Use PHP v8 nullsafe operator

## [v6.0.3] - 2022-01-02
- Update str global helpers to check for empty strings

## [v6.0.2] - 2021-12-20
- Add two new global helpers get_str_contains & get_str_endswith

## [v6.0.1] - 2021-10-05
- Refactor timezone getter/setter and fix related issues (fix #57)

## [v6.0.0] - 2021-08-22
- Drop PHP v7 support, and upgrade rinvex package dependencies to next major version
- Update composer dependencies

## [v5.0.19] - 2021-08-06
- Move application layer helpers to cortex/foundation module helpers (43bf9ad)
- Refactor accessarea obscure and indexable config options (7891a61)
- Refactor HashidsTrait ID obscuration (2b4125b)
- Update composer dependency codedungeon/phpunit-result-printer (08d9601)
- Rename areas to accessareas (8106779)

## [v5.0.18] - 2021-06-19
- Add array_diff_assoc_recursive global helper
- Allow modules to override core packages rules
- Update HasTranslations.php (#37)

## [v5.0.17] - 2021-05-26
- Set default status code to null, we'll handle it!

## [v5.0.16] - 2021-05-26
- Fix compatibility issue with diglactic/laravel-breadcrumbs V7

## [v5.0.15] - 2021-05-25
- Gracefully handle unset array identifiers

## [v5.0.14] - 2021-05-24
- Refactor `intend` function to return more accurate HTTP status codes

## [v5.0.13] - 2021-05-11
- Update spatie/laravel-schemaless-attributes composer dependency

## [v5.0.12] - 2021-05-04
- Catch static & non-static calls for static and non-static methods before calling macros

## [v5.0.11] - 2021-04-27
- Revert custom package to default felixkiss/uniquewith-validator
- Fix str_replace string conversion issue

## [v5.0.10] - 2021-03-15
- Fix timezone issue with created_at, updated_at, and automatically generated timestamp fields

## [v5.0.9] - 2021-03-15
- Revert "Fix user timezone conversion on read/update - ignore current user timezone when using setAttribute (#47)"

## [v5.0.8] - 2021-02-28
- Fix user timezone conversion on read/update - ignore current user timezone when using setAttribute (#47)

## [v5.0.7] - 2021-02-24
- Simplify and utilize request()->user() and request()->guard()
- Simplify and utilize request()->accessarea()
- Fix wrong http status code 0 on redirect response
- Add `unhashId` method to unhash hashids per model
- Refactor `intend` global helper to support API response better
- Move `get_access_area` global helper to `Request::getAccessArea` method
- Add `similar` collection macro

## [v5.0.6] - 2021-02-16
- Fix macroable conflict with dynamic relationships
- Refactor `get_access_area` global helper to support all possible scenarios:
    1. Route matched and is an accessarea request (ex. /adminarea/users)
    2. Route matched and is an API request (ex. /api/users)
    3. Catch other use cases:
       3.1. Route NOT matched / Wrong URL (ex. 404 error)
       3.2. Route matched but NOT a valid accessarea (could happen if route is mistakenly named, make sure route names contain valid accessarea prefix)

## [v5.0.5] - 2021-02-11
- Add ability for artisan commands to validate answers
- Add `get_access_area` helper function

## [v5.0.4] - 2021-02-06
- Simplify service provider model registration into IoC
- Enforce consistency

## [v5.0.3] - 2020-12-25
- Fix wrong composer dependency version constraints

## [v5.0.2] - 2020-12-25
- Switch outdated composer dependencies to temporary compatible forks until it's updated

## [v5.0.1] - 2020-12-25
- Add support for PHP v8

## [v5.0.0] - 2020-12-22
- Upgrade to Laravel v8

## [v4.3.1] - 2020-07-16
- Refactor timezone implementation to fix read/write date attributes

## [v4.3.0] - 2020-07-16

src/Support/helpers.php
<?php

declare(strict_types=1);

if (! function_exists('mimetypes')) {
    /**
     * Get valid mime types.
     *
     * @see https://github.com/symfony/http-foundation/blob/3.0/File/MimeType/MimeTypeExtensionGuesser.php
     * @see http://svn.apache.org/repos/asf/httpd/httpd/trunk/docs/conf/mime.types
     *
     * @return array
     */
    function mimetypes()
    {
        return json_decode(file_get_contents(__DIR__.'/../../resources/data/mimetypes.json'), true);
    }
}

if (! function_exists('timezones')) {
    /**
     * Get valid timezones.
     *
     * @return array
     */
    function timezones()
    {
        return array_combine(timezone_identifiers_list(), timezone_identifiers_list());
    }
}

if (! function_exists('timeoffsets')) {
    /**
     * Get valid time offsets.
     *
     * @return array
     */
    function timeoffsets()
    {
        return [
            '-1200' => 'UTC -12:00',
            '-1100' => 'UTC -11:00',
            '-1000' => 'UTC -10:00',
            '-0930' => 'UTC -09:30',
            '-0900' => 'UTC -09:00',
            '-0800' => 'UTC -08:00',
            '-0700' => 'UTC -07:00',
            '-0600' => 'UTC -06:00',
            '-0500' => 'UTC -05:00',
            '-0400' => 'UTC -04:00',
            '-0330' => 'UTC -03:30',
            '-0300' => 'UTC -03:00',
            '-0200' => 'UTC -02:00',
            '-0100' => 'UTC -01:00',
            '+0000' => 'UTC ±00:00',
            '+0100' => 'UTC +01:00',
            '+0200' => 'UTC +02:00',
            '+0300' => 'UTC +03:00',
            '+0330' => 'UTC +03:30',
            '+0400' => 'UTC +04:00',
            '+0430' => 'UTC +04:30',
            '+0500' => 'UTC +05:00',
            '+0530' => 'UTC +05:30',
            '+0545' => 'UTC +05:45',
            '+0600' => 'UTC +06:00',
            '+0630' => 'UTC +06:30',
            '+0700' => 'UTC +07:00',
            '+0800' => 'UTC +08:00',
            '+0845' => 'UTC +08:45',
            '+0900' => 'UTC +09:00',
            '+0930' => 'UTC +09:30',
            '+1000' => 'UTC +10:00',
            '+1030' => 'UTC +10:30',
            '+1100' => 'UTC +11:00',
            '+1200' => 'UTC +12:00',
            '+1245' => 'UTC +12:45',
            '+1300' => 'UTC +13:00',
            '+1400' => 'UTC +14:00',
        ];
    }
}

if (! function_exists('array_search_recursive')) {
    /**
     * Recursively searches the array for a given value and returns the corresponding key if successful.
     *
     * @param mixed $needle
     * @param array $haystack
     *
     * @return mixed
     */
    function array_search_recursive($needle, $haystack)
    {
        foreach ($haystack as $key => $value) {
            $current_key = $key;
            if ($needle === $value || (is_array($value) && array_search_recursive($needle, $value) !== false)) {
                return $current_key;
            }
        }

        return false;
    }
}

if (! function_exists('array_trim_recursive')) {
    /**
     * Recursively trim elements of the given array.
     *
     * @param mixed  $values
     * @param string $charlist
     *
     * @return mixed
     */
    function array_trim_recursive($values, $charlist = " \t\n\r\0\x0B")
    {
        if (is_array($values)) {
            return array_map('array_trim_recursive', $values);
        }

        return is_string($values) ? trim($values, $charlist) : $values;
    }
}

if (! function_exists('array_diff_assoc_recursive')) {
    /**
     * Computes the recursive difference of arrays with additional index check.
     *
     * @param array $array1
     * @param array $array2
     * @param bool  $onlyDiff
     *
     * @return array
     */
    function array_diff_assoc_recursive(array $array1, array $array2, bool $onlyDiff = true)
    {
        $difference = [];

        foreach ($array1 as $key => $value) {
            if (is_array($value)) {
                if (! isset($array2[$key]) || ! is_array($array2[$key])) {
                    $difference[$key] = $value;
                } elseif (! empty($subDiff = array_diff_assoc_recursive($value, $array2[$key]))) {
                    $difference[$key] = $onlyDiff ? $subDiff : array_merge($array2[$key], $subDiff);
                }
            } elseif (! array_key_exists($key, $array2) || $array2[$key] !== $value) {
                $difference[$key] = $value;
            }
        }

        return $difference;
    }
}

if (! function_exists('get_str_contains')) {
    /**
     * Get str contains the given string.
     *
     * @param string          $haystack
     * @param string|string[] $needles
     *
     * @return bool
     */
    function get_str_contains($haystack, $needles)
    {
        foreach ((array) $needles as $needle) {
            if (! empty($needle) && mb_strpos($haystack, $needle) !== false) {
                return $needle;
            }
        }

        return false;
    }
}

if (! function_exists('get_str_endswith')) {
    /**
     * Get str ends with the given string.
     *
     * @param string          $haystack
     * @param string|string[] $needles
     *
     * @return bool
     */
    function get_str_endswith($haystack, $needles)
    {
        foreach ((array) $needles as $needle) {
            if (! empty($needle) && mb_substr($haystack, -mb_strlen($needle)) === (string) $needle) {
                return $needle;
            }
        }

        return false;
    }
}
